src/lib/TopLevel.hs

```diff
@@ -427,61 +521,129 @@ addResultCtx :: SourceBlock -> Result -> Result
 addResultCtx block (Result outs errs) =
   Result outs (addSrcTextContext (sbOffset block) (sbText block) errs)
 
-logTop :: MonadPasses m => Output -> m n ()
+logTop :: TopLogger m => Output -> m ()
 logTop x = logIO [x]
 
-findModulePath :: MonadInterblock m => ModuleName -> m FilePath
-findModulePath moduleName = do
-  let fname = moduleName ++ ".dx"
-  specifiedPath <- libPath <$> getConfig
-  case specifiedPath of
-    Nothing -> liftIO $ getDataFileName $ "lib/" ++ fname
-    Just path -> return $ path </> fname
-
--- === instances ===
-
-instance ConfigReader (PassesM n) where
-  getConfig = PassesM $ asks \(_, _, cfg) -> cfg
+logPass :: Topper m => Pretty a => PassName -> m n a -> m n a
+logPass passName cont = do
+  logTop $ PassInfo passName $ "=== " <> pprint passName <> " ==="
+  logTop $ MiscLog $ "Starting "++ pprint passName
+  result <- cont
+  logTop $ PassInfo passName $ "=== Result ===\n" <> pprint result
+  return result
 
-instance TopBuilder PassesM where
+loadModuleSource :: MonadIO m => EvalConfig -> ModuleSourceName -> m File
+loadModuleSource config moduleName = do
+  fnameStem <- case moduleName of
+    OrdinaryModule moduleName' -> return moduleName'
+    Prelude -> return "prelude"
+    Main -> error "shouldn't be trying to load the source for main"
+  let fname = fnameStem ++ ".dx"
+  fullPath <- case libPath config of
+      Nothing -> liftIO $ getDataFileName $ "lib/" ++ fname
+      Just path -> return $ path </> fname
+  readFileWithHash fullPath
+
+-- === saving cache to disk ===
+
+-- None of this is safe in the presence of multiple processes trying to interact
+-- with the cache. But we plan to fix that by using an actual database.
+
+loadCache :: MonadIO m => m TopStateEx
+loadCache = liftIO do
+  cachePath <- getCachePath
+  cacheExists <- doesFileExist cachePath
+  if cacheExists
+    then do
+      decoded <- decode <$> BS.readFile cachePath
+      case decoded of
+        Right result -> restorePtrSnapshots result
+        _            -> removeFile cachePath >> return initTopState
```

Comment:
Warn? That shouldn't really happen

Reply:
Good point. This is just copied over from the old cache code. This should fail loudly.

Reply:
Well perhaps not fail, but dropping the cache entirely is costly and might warrant a warning.

Reply:
If we can't even decode the store, something has gone wrong (probably version skew). So I think an error actually makes sense.